### PR TITLE
feat(DIN-27): spell slot tracking and display

### PR DESCRIPTION
**Pull request by dinokong0128**
## Summary

- **`SPELL_SLOTS_BY_CLASS_LEVEL`** lookup table in `backend/utils/dnd.py` covering levels 1–5 for all caster classes (full, half, and Warlock pact magic); non-casters map to `None`
- **`upsert_player`** now populates `stats.spell_slots` (`{"1": {"max": N, "used": 0}, ...}`) for spellcasting classes and `null` for non-casters — no DB migration required (JSONB)
- **`apply_state_changes()`** handles two new `<state_changes>` keys: `spell_slot_use` (increment used, cap at max) and `spell_slots_recharge` (reset all used to 0 on long rest)
- **System prompt** party line now includes slot remaining counts, spell save DC, spell attack bonus, and cantrips so Claude can respect slot availability
- **`CharacterSheetPanel`** renders a Spell Slots section (hidden for non-casters) with pip display (● used / ○ available), count labels, and cantrips list
- **TypeScript type** for `PlayerRow.stats` extended with `spell_slots?` and `cantrips?`

## Test plan

- [ ] 19 new tests in `backend/tests/test_dnd.py` — SPELL_SLOTS_BY_CLASS_LEVEL structure + `get_spell_slots` edge cases
- [ ] 4 new tests in `test_dm_service.py` — `spell_slot_use` increments, cap at max, `spell_slots_recharge` reset, best-effort on DB failure
- [ ] 2 new tests in `test_players_inventory.py` — Wizard upsert includes spell_slots, Fighter upsert has null
- [ ] 3 new tests in `test_dm_tasks.py` — spell slots/cantrips appear in system prompt for Wizard, absent for Fighter
- [ ] 6 new tests in `CharacterSheetPanel.spell.test.tsx` — section visibility, pip counts, count label, hidden zero-max levels, cantrips
- [ ] `npx tsc --noEmit` passes (0 errors)
- [ ] Full suites: 202 backend + 505 frontend tests all green

https://claude.ai/code/session_0126Ds6pbRKUMTUWz72fxuZk

### CODE REVIEW

File: backend/tasks/dm_tasks.py [hunk 1]
```diff
@@ -167,6 +167,41 @@ def _ability_mod(score: int) -> str:
                 f"  Equipment: {items}"
             )
 
+            # Spell slots (DIN-27)
+            spell_slots = stats.get("spell_slots")
+            if spell_slots:
+                _ordinals = {1: "1st", 2: "2nd", 3: "3rd", 4: "4th", 5: "5th"}
+                slot_parts = []
+                for lvl_key in sorted(spell_slots.keys(), key=lambda x: int(x)):
```
[the remark this answers is not in the document]
Comment (dinokong0128) [reply backend/tasks/dm_tasks.py]: Fixed in dd3e0df. Changed the sort to filter keys first with `(k for k in spell_slots.keys() if isinstance(k, str) and k.isdigit())` before passing to `sorted(..., key=int)`. Non-numeric keys are silently skipped rather than raising `ValueError`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: backend/api/routes/players.py [hunk 2]
```diff
@@ -108,6 +108,15 @@ async def upsert_player(
...
     hp_max = calculate_hp_max(body.character_class, body.stats.con)
     stats_dict = body.stats.model_dump(by_alias=True)
+
+    raw_slots = get_spell_slots(body.character_class, body.level)
+    if raw_slots is None:
+        stats_dict["spell_slots"] = None
+    else:
+        stats_dict["spell_slots"] = {
+            str(lvl): {"max": raw_slots.get(lvl, 0), "used": 0}
+            for lvl in range(1, 4)
```
[the remark this answers is not in the document]
Comment (dinokong0128) [reply backend/api/routes/players.py:119]: Fixed in dd3e0df. Changed `if raw_slots is None` to `if not raw_slots`, so both `None` (non-casters) and `{}` (casters with no slots at the current level, e.g. Paladin/Ranger level 1) store `null`. Downstream truthiness checks now correctly treat these characters as non-spellcasting until they gain slots.

---
_Generated by [Claude Code](https://claude.ai/code)_
